File: projetos-fenix-para-estudo/t2ti-erp-fenix-cadastros-completo/fontes/backend/node/t2ti_erp_fenix/src/service/banco_service.js
conexao = require('../config/db_config.js');
const Banco = require('../model/banco.js');

class BancoService {

    static ConsultarLista(result) {
        Banco.findAll().then(listaBanco => {
            result(listaBanco, null);
        }).catch(function (erro) {
            result(null, erro);
        });
    }

    static ConsultarListaFiltroValor(filtro, result) {
        var sql = "select * from banco where " + filtro.where;
        Banco.sequelize
            .query(sql, {
                model: Banco,
                mapToModel: true // para verificar os campos mapeados
            })
            .then(listaBanco => {
                result(listaBanco, null);
            })
            .catch(function (erro) {
                result(null, erro);
            });
    }

    static ConsultarObjeto(Id, result) {
        Banco.findByPk(Id).then(banco => {
            result(banco, null);
        }).catch(function (erro) {
            result(null, erro);
        });
    }

    static Inserir(bancoJson, result) {
        Banco.create(bancoJson).then(banco => {
            result(banco, null);
        }).catch(function (erro) {
            result(null, erro);
        });
    }

    static Alterar(bancoJson, result) {
        Banco.update(bancoJson, {
            where: {
                id: bancoJson.id
            }
        }).then(registrosAtualizados => {
            result(registrosAtualizados[0], null);
        }).catch(function (erro) {
            result(null, erro);
        });
    }

    static Excluir(Id, result) {
        Banco.destroy({
            where: {
                id: Id
            }
        }).then(registrosAtualizados => {
            result(registrosAtualizados[0], null);
        }).catch(function (erro) {
            result(null, erro);
        });
    }

}

module.exports = BancoService;

File: projetos-fenix-para-estudo/t2ti-erp-fenix-servidores-provider-inicio/fontes/backend/delphi/dmvc/Service/BancoService.pas
unit BancoService;

interface

uses
  System.SysUtils, System.Generics.Collections, ServiceBase, Banco;

type

  TBancoService = class(TServiceBase)
  public
    class function ConsultarLista: TObjectList<TBanco>;
    class function ConsultarListaFiltroValor(campo: string; valor: string): TObjectList<TBanco>;
    class function ConsultarObjeto(const AId: Integer): TBanco;
    class procedure Inserir(ABanco: TBanco);
    class procedure Alterar(ABanco: TBanco);
    class procedure Excluir(ABanco: TBanco);
  end;

const
  INSERT = 'INSERT INTO BANCO '+
           '(ID, CODIGO, NOME, URL) '+
           'VALUES (:NEW_ID, :NEW_CODIGO, :NEW_NOME, :NEW_URL)';
  UPDATE = 'UPDATE BANCO '+
           'SET ID = :NEW_ID, CODIGO = :NEW_CODIGO, NOME = :NEW_NOME, '+
           'URL = :NEW_URL '+
           'WHERE ID = :OLD_ID';
  DELETE = 'DELETE FROM BANCO WHERE ID = :OLD_ID';


implementation

uses
  FireDAC.Stan.Option, FireDAC.Comp.Client, FireDAC.Stan.Param,
  MVCFramework.FireDAC.Utils, MVCFramework.DataSet.Utils,
  MVCFramework.Serializer.Commons, MVCFramework.Commons;

{ TBancoService }

class function TBancoService.ConsultarLista: TObjectList<TBanco>;
begin
  Query := TFDQuery.Create(nil);
  Query.Connection := GetDataModule.Conexao;
  try
    Query.Open('SELECT * FROM BANCO ORDER BY ID');
    Result := Query.AsObjectList<TBanco>;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TBancoService.ConsultarListaFiltroValor(campo, valor: string): TObjectList<TBanco>;
begin
  Query := TFDQuery.Create(nil);
  Query.Connection := GetDataModule.Conexao;
  try
    Query.Open('SELECT * FROM BANCO where ' + campo + ' like "%' + valor + '%"');
    Result := Query.AsObjectList<TBanco>;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class function TBancoService.ConsultarObjeto(const AId: Integer): TBanco;
begin
  Query := TFDQuery.Create(nil);
  Query.Connection := GetDataModule.Conexao;
  try
    Query.Open('SELECT * FROM BANCO WHERE ID = :ID', [AId]);
    if not Query.Eof then
      Result := Query.AsObject<TBanco>
    else
      Result := nil;
  finally
    Query.Close;
    Query.Free;
  end;
end;

class procedure TBancoService.Inserir(ABanco: TBanco);
var
  Cmd: TFDCustomCommand;
begin
  ABanco.ValidarInsercao;

  UpdateSQL := TFDUpdateSQL.Create(nil);
  UpdateSQL.Connection := GetDataModule.Conexao;
  UpdateSQL.InsertSQL.Add(INSERT);
  try
    Cmd := UpdateSQL.Commands[arInsert];
    TFireDACUtils.ObjectToParameters(Cmd.Params, ABanco, 'NEW_');
    Cmd.Execute;
  finally
    UpdateSQL.Free;
  end;
end;

class procedure TBancoService.Alterar(ABanco: TBanco);
var
  Cmd: TFDCustomCommand;
begin
  ABanco.ValidarAlteracao;

  UpdateSQL := TFDUpdateSQL.Create(nil);
  UpdateSQL.Connection := GetDataModule.Conexao;
  UpdateSQL.ModifySQL.Add(UPDATE);
  try
    Cmd := UpdateSQL.Commands[arUpdate];
    TFireDACUtils.ObjectToParameters(Cmd.Params, ABanco, 'NEW_');
    Cmd.ParamByName('OLD_ID').AsInteger := ABanco.Id;
    Cmd.Execute;
    if Cmd.RowsAffected <> 1 then
      raise EMVCException.Create('Objeto com ID inválido [Alterar Banco]', '', 0, 400);
  finally
    UpdateSQL.Free;
  end;
end;

class procedure TBancoService.Excluir(ABanco: TBanco);
var
  Cmd: TFDCustomCommand;
begin
  ABanco.ValidarExclusao;

  UpdateSQL := TFDUpdateSQL.Create(nil);
  UpdateSQL.Connection := GetDataModule.Conexao;
  UpdateSQL.DeleteSQL.Add(DELETE);
  try
    Cmd := UpdateSQL.Commands[arDelete];
    TFireDACUtils.ObjectToParameters(Cmd.Params, ABanco, 'OLD_');
    Cmd.Execute;
  finally
    UpdateSQL.Free;
  end;
end;

end.
